src/tls.rs

```diff
@@ -1,6 +1,234 @@
+        // ESP docs treat 0 as error, but in Rust it's common to return 0 from `Read::read` to indicate eof
+        if ret >= 0 {
+            Ok(ret as usize)
+        } else {
+            Err(io::Error::new(io::ErrorKind::Other, Error::ReadError(ret)))
+        }
+    }
+}
+
+impl Write for EspTls {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        // cannot call esp_tls_conn_write bc it's inline
+        //let ret = unsafe { esp_idf_sys::esp_tls_conn_write(self.raw, buf.as_ptr(), buf.len()) };
+        let esp_tls = unsafe { std::ptr::read_unaligned(self.raw) };
+        let write_func = esp_tls.write.unwrap();
+        let ret = unsafe { write_func(self.raw, buf.as_ptr() as *const i8, buf.len()) } as i32;
+        if ret >= 0 {
+            Ok(ret as usize)
+        } else {
+            Err(io::Error::new(io::ErrorKind::Other, Error::WriteError(ret)))
+        }
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        Ok(())
+    }
+}
+
+impl Drop for EspTls {
+    fn drop(&mut self) {
+        unsafe {
+            esp_idf_sys::esp_tls_conn_destroy(self.raw);
+        }
+    }
+}
+
+pub struct Config<'a> {
+    pub alpn_protos: Option<&'a [&'a CStr]>,
+    pub ca_cert: Option<X509<'a>>,
+    pub client_cert: Option<X509<'a>>,
+    pub client_key: Option<X509<'a>>,
+    pub client_key_password: Option<&'a CStr>,
+    pub non_block: bool,
+    pub use_secure_element: bool,
+    pub timeout_ms: u32,
+    pub use_global_ca_store: bool,
+    pub common_name: &'a CStr,
+    pub skip_common_name: bool,
+    pub keep_alive_cfg: Option<KeepAliveConfig>,
+    pub psk_hint_key: Option<PskHintKey<'a>>,
+    // TODO crt_bundle_attach
+    // TODO ds_data
+    pub is_plain_tcp: bool,
+    pub if_name: esp_idf_sys::ifreq,
+}
+
+impl<'a> Default for Config<'a> {
+    fn default() -> Self {
+        Self {
+            alpn_protos: Default::default(),
+            ca_cert: None,
+            client_cert: None,
+            client_key: None,
+            client_key_password: Default::default(),
+            non_block: Default::default(),
+            use_secure_element: Default::default(),
+            timeout_ms: Default::default(),
+            use_global_ca_store: Default::default(),
+            common_name: Default::default(),
+            skip_common_name: Default::default(),
+            keep_alive_cfg: Default::default(),
+            psk_hint_key: Default::default(),
+            is_plain_tcp: Default::default(),
+            if_name: Default::default(),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct KeepAliveConfig {
+    /// Enable keep-alive timeout
+    pub enable: bool,
+    /// Keep-alive idle time (second)
+    pub idle: Duration,
+    /// Keep-alive interval time (second)
+    pub interval: Duration,
+    /// Keep-alive packet retry send count
+    pub count: u32,
+}
+
+pub struct PskHintKey<'a> {
+    pub key: &'a [u8],
+    pub hint: &'a CStr,
+}
+
+#[derive(Clone, Debug)]
```

Comment:
As per above, please do not introduce a custom error type. Specifically when implementing the `embedded_svc::io` traits, you can plug in there `EspIOError`, which is a wrapper around `EspError`.